### PR TITLE
(Summer 2022)Feat add golang runtime metrics

### PR DESCRIPTION
**Pull request by easonyipj**
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [x] If it's UI related, attach the screenshots below.
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

# Feature Description
This feature is about the golang runtime metrics collcection, while using go2Sky gRPCReporter, it can collect golang runtime metrics automatically, and sending the metrics to OAP Server where you can monitor these metrics in general-Instance dashboard.

Supported metrics:

- instance_golang_heap_alloc
bytes of allocated heap objects.
- instance_golang_stack_used
bytes in stack spans.
- instance_golang_gc_count
the gc count in one minute
- instance_golang_gc_pause_time
the sum of all gc pause time in one minute
- instance_golang_live_goroutines_count
the number of goroutines that currently exist.
- instance_golang_os_threads_count
the number of records in the thread creation profile
- instance_host_cpu_used_rate
the percentage of cpu used combined.
- instance_host_mem_used_rate
the memory used rate

# How it works
1. The skywalking-meter-receiver-plugin will receive the meter data sending by go2sky
2. The meter sysytem will process these meters according to the meter-analyzer-config/golang.yaml
3. The meter panel defined in general-instance.json, the ui system will load there panels when you open the General-Instance dashboard and choose the Golang tap

# How to use
1. Update application.yml to add golang in meterAnalyzerActiveFiles
`meterAnalyzerActiveFiles: ${SW_METER_ANALYZER_ACTIVE_FILES:datasource,threadpool,satellite,golang}`
3. Start your go2sky agent and oap server
4. Open the  General-Instance dashboard and choose Golang tap

# Example
example image

### CODE REVIEW

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: Is this a gauge?

**Comment by wu-sheng** [comment]
Also, if you want to activate this, you should add this file to the activated in the application.yml.

**Comment by easonyipj** [comment]
> Also, if you want to activate this, you should add this file to the activated in the application.yml.

Yes, I just added that

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json:1107]: Pending on https://github.com/apache/skywalking-booster-ui/pull/164

**Comment by wu-sheng** [comment]
> > Also, if you want to activate this, you should add this file to the activated in the application.yml.
> 
> Yes, I just added that

Locally? I can't see it from this PR.

**Comment by easonyipj** [comment]
> > > Also, if you want to activate this, you should add this file to the activated in the application.yml.
> > 
> > 
> > Yes, I just added that
> 
> Locally? I can't see it from this PR.

Sorry, I didn't make it clear, I mean I have mentioned the user who wants to see golang meter should active the golang.yaml in application.yml on the document. Should I add golang to meterAnalyzerActiveFiles in this pr directly neither than let user to active?

**Comment by wu-sheng** [comment]
> I mean I have mentioned the user who wants to see golang meter should active the golang.yaml in application.yml on the document.

Which document? I didn't notice that too.

> Should I add golang to meterAnalyzerActiveFiles in this pr directly neither than let user to active?

Yes, you should.

**Comment by easonyipj** [comment]
> > I mean I have mentioned the user who wants to see golang meter should active the golang.yaml in application.yml on the document.
> 
> Which document? I didn't notice that too.
> 
> > Should I add golang to meterAnalyzerActiveFiles in this pr directly neither than let user to active?
> 
> Yes, you should.

The pr docs, I mentioned it at the first step on "How to use" module, I will directly add golang to meterAnalyzerActiveFiles in application.yml, thanks for reminding

**Comment by wu-sheng** [comment]
> The pr docs, I mentioned it at the first step on "How to use" module

Could you share the link here? I can't follow your point.

**Comment by wu-sheng** [comment]
image

This is all I can see about the changes.

**Comment by easonyipj** [comment]
> Don't remove the template, which indicates you should update `changes.md`.

OK， I will change it

**Comment by easonyipj** [comment]
> image
> 
> This is all I can see about the changes.

Sorry, I mean this, maybe it should not be called "pr docs"? And I will change there words follow the pr template
example image

**Comment by wu-sheng** [comment]
No, that is not docs. And there is no PR's docs thing. 

This is the [official doc](https://skywalking.apache.org/docs/#SkyWalking), which is built from [here](https://github.com/apache/skywalking/blob/master/docs/README.md) from the latest to every selected version.

What you wrote here is called the content or description of this PR. The purpose of this is to introduce what you changed and why you changed in the way. You should write what you had written here, that is good practice. Just it isn't being called a doc.

**Comment by easonyipj** [comment]
> No, that is not docs. And there is no PR's docs thing.
> 
> This is the [official doc](https://skywalking.apache.org/docs/#SkyWalking), which is built from [here](https://github.com/apache/skywalking/blob/master/docs/README.md) from the latest to every selected version.
> 
> What you wrote here is called the content or description of this PR. The purpose of this is to introduce what you changed and why you changed in the way. You should write what you had written here, that is good practice. Just it isn't being called a doc.

Thank you for your correction！

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json:1107]: According to UI side PR, this should be udpated.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: instance_golang_stack represents the bytes in stack spans right now
instance_golang_gc_time represents  the most recent gc pause time
those are MeterSingleValue?

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: My point is about GC time. How a metric could be called recent? 
What is the time duration about the recent?

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: > 
time duration is decided by user input, if user input is 15 seconds, than every 15 seconds will collect the latest gc's puase time using

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: Then, what is your set? As the GC time is splitted, you should use sum function to merge, and use sum downsampling for hour and day level metric aggregation.

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: The default downsampling is an avg, check hour level GC time you set now, you would see a hour GC time is close to a minute GC time.

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: OK, Should I change to "instance_golang_gc_time.downsampling(SUM)" ?

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/golang.yaml:24]: OK, thanks, I will test this method

File: oap-server/server-starter/src/main/resources/application.yml
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/application.yml:237]:
```suggestion
    meterAnalyzerActiveFiles: ${SW_METER_ANALYZER_ACTIVE_FILES:datasource,threadpool,satellite,go-runtime} # Which files could be meter analyzed, files split by ","
```

Please update the MAL file accordingly.

File: docs/en/changes/changes.md
Comment (wu-sheng) [inline docs/en/changes/changes.md:31]:
```suggestion
* Support Golang runtime metrics analysis.
```

File: oap-server/server-starter/src/main/resources/application.yml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/application.yml:237]: Ok, done.

**Comment by wu-sheng** [comment]
Pending your metrics verification. https://github.com/apache/skywalking/pull/9711 should get approval first.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:34]: Actually, these are host meters, I will change these.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: I had tested, and the new instance_golang_gc_time  meter in UI represents the sum of all gc puase time that collected in a one minute. For example, if the meter interval is 15s, then go2sky will collect gc_time four times in a minute, and UI will show the sum of four collected gc_pause_time, but gc will happened more than 4 time in one minute(around 500 times one minute),
so these sum  of incomplete gc pause time may confuse users?

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: What do you mean? I am not following. The sum of time cost of GC should be reasonable, no matter how many time you collected and reported.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json:1120]:
```suggestion
                    "title": "Golang GC Count Per Minute"
```

This seems actually the real meaning of this metric if you report this per minute.
In hour/day level, the gc_count(increase) would still get `avg` downsampling. We always get a `count/min`

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/general/general-instance.json:1120]: Report period matters in the MAL analysis case, as you are reporting statistical data. So, I recommend you either report this per minute or not reported it.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: I mean the ui only shows the sum of time cost of **collected** GC (only four times in one minute if the interval is 15s) not the **all** GC(around 500 times in one minute) in one minute, so if ui shows the gc_cost_time is 0.5ms at 20:01, actually, the total gc_pause_time would more than 0.5ms at 20:01.  I just confuse about that, thanks for helping me.

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: Which type actually this metric is? Is it a counter or gauge? If it is a gauge, then it should always report value from last time, the report period should not matter.

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: The current collection method is a gauge, Because it just collect the latest single gc's pause time, I think I should change the collection method to collect the total gc_pause_time in the collect period, thanks.

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: Yes, so, you could check whether there is a hook/callback to measure this. Reading GC pause time periodly seems a back practice.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:19]: > I think we should add an e2e to test these metrics?

Ok，should the e2e be put in go2sky?

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: I use stats.PauseTotal to collect the total gc time which represents the sum of gc time after the instance started. So it is a counter meter. And I use increase('PT1M') to calculate the sum of one minute. So the metrics is instance_golang_gc_time.increase('PT1M'). I have tested that and it worked, is this reasonable?

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (wankai123) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:19]: yes, you can put it in go2sky.

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:19]: > yes, you can put it in go2sky.

OK

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: Yes, then this is the GC pause time per minute metric.

Also, please try to re-word all the metrics names, they should be more accurate, formal and clear.

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:24]: > Yes, then this is the GC pause time per minute metric.
> 
> Also, please try to re-word all the metrics names, they should be more accurate, formal and clear.

Ok, thanks, I will do this work.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:34]: Yes, I am doing some changes on go2sky, and I will change these metrics names later, and pushed them after tested.

File: oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:30]: I think `thread_num` and `goroutine_num` are good, count usually indicates it is a counter to do `rate` or `increase`.

Comment (easonyipj) [reply oap-server/server-starter/src/main/resources/meter-analyzer-config/go-runtime.yaml:30]: OK, i will change these to _num_

**Comment by wu-sheng** [comment]
Also, you need to sync with the master branch, and resolve conflicts.(just a changelog doc)

**Comment by wu-sheng** [comment]
I think this PR generally is good, we just wait for your agent side ready.

**Comment by easonyipj** [comment]
> I think this PR generally is good, we just wait for your agent side ready.

Ok, thanks for your guidance, I will add e2e test on go2sky, and then test it before the agent side ready

**Comment by wu-sheng** [comment]
You have to test manually as if this PR doesn't get merged, you can't find the image to test.

@arugal Do we use agent test tool in go2sky repo?

**Comment by arugal** [comment]
> 

Yes, just like our plugin e2e test https://github.com/SkyAPM/go2sky/blob/master/plugins/http/test/docker-compose.yml

**Comment by wu-sheng** [comment]
OK, @easonyipj then agent side could be tested without backend OAP in the e2e actually.

The server side would not be able to test.

@arugal Do you prefer to add e2e to skywalking main repo, which could be a little too heavy? Or we should have a real OAP placed e2e for this case? Or we could leave backend OAP manually tested only?

**Comment by easonyipj** [comment]
> OK, @easonyipj then agent side could be tested without backend OAP in the e2e actually.
> 
> The server side would not be able to test.
> 
> @arugal Do you prefer to add e2e to skywalking main repo, which could be a little too heavy? Or we should have a real OAP placed e2e for this case? Or we could leave backend OAP manually tested only?

OK, I will do this on agent side only. Thanks for you all.

**Comment by wu-sheng** [comment]
@easonyipj Let me explain a little more about this

> Yes, just like our plugin e2e test https://github.com/SkyAPM/go2sky/blob/master/plugins/http/test/docker-compose.yml

The docker compose is using `ghcr.io/apache/skywalking-agent-test-tool/mock-collector` as the backend rather than OAP server to test. As a result, all MAL scripts are not going to be tested if you ref this case to test.

Mock collector focuses on testing the data agent reported, whether it matches the expectation, but not how they would be analyzed.

**Comment by easonyipj** [comment]
> 

OK, thanks very much, I will learn how to use these to build an e2e test.

**Comment by arugal** [comment]
@wu-sheng https://github.com/SkyAPM/go2sky/pull/168 is merged

**Comment by easonyipj** [comment]
> @wu-sheng [SkyAPM/go2sky#168](https://github.com/SkyAPM/go2sky/pull/168) is merged
> 
> Hi @easonyipj After this pr merged, please use oap as the server to add e2e test in go2sky

OK, got it, thanks

**Comment by wu-sheng** [comment]
Thanks for the update, so, this PR would still be hold still e2e added, right?

**Comment by arugal** [comment]
> Thanks for the update, so, this PR would still be hold still e2e added, right?

need to add the e2e test in the main repo？

**Comment by wu-sheng** [comment]
No, we don't require that. I mean should I need the go2sky side e2e added? If you have tested this and feel OK to merge, I am good.

**Comment by arugal** [comment]
> If you have tested this and feel OK to merge, I am good.

Yes, we encountered some problems when using[ skywalk-agent-test-tool](https://github.com/apache/skywalking-agent-test-tool) for e2e test, we will add go2sky e2e testing after this pr merged

**Review by wu-sheng — APPROVED** [review]

**Comment by wu-sheng** [comment]
Merged.

**Pull request merged**
